### PR TITLE
Improve Javadoc to show optional/required properties and include sample codesnippets

### PR DESCRIPTION
**Pull request by haolingdong-msft**
Fix https://github.com/Azure/autorest.java/issues/1464
Required/Optional flag:
image

Query table description:
image

Header table description:
image

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java [hunk 1]
```diff
@@ -55,16 +56,21 @@ protected static void generateProtocolMethodJavadoc(ClientMethod clientMethod, J
                     .collect(Collectors.toList());
             if (!headerParameters.isEmpty()) {
                 optionalParametersJavadoc("Header Parameters", headerParameters, commentBlock);
+                commentBlock.line("You can add these to a request with {@code new RequestOptions.addHeader()}");
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java:59]: Should we use `setHeader()`?

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java [hunk 2]
```diff
@@ -247,4 +253,8 @@ private static String methodParameterDescriptionOrDefault(ClientMethodParameter
         }
         return doc;
     }
+
+    private static String appendOptionalOrRequiredAttribute(boolean isRequired, boolean isRootSchema) {
+        return isRootSchema ? "" : isRequired ? " (Required)" : " (Optional)";
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java:258]: nit, not sure upper-case or lower-case be better, just for discussion.

File: protocol-tests/swagger/required-optional-body.json [hunk 3]
```diff
@@ -0,0 +1,134 @@
+{
```
Comment (weidongxu-microsoft) [inline protocol-tests/swagger/required-optional-body.json:1]: Does this test provide more scenario than other existing tests (testserver and others), or can it be combined with existing?

We might not want to add too many test swagger, as it have some cost on dev time and CI time.

File: protocol-resilience-test/llcupdate1/src/main/java/fixtures/llcresi/DpgClient.java [hunk 4]
```diff
@@ -125,7 +131,7 @@ public Response<BinaryData> putRequiredOptionalWithResponse(String requiredParam
     }
 
     /**
-     * POST a JSON or a JPEG.
+     * POST a JSON or a JPEG. You can add these to a request with {@code new RequestOptions.addHeader()}
```
Comment (weidongxu-microsoft) [inline protocol-resilience-test/llcupdate1/src/main/java/fixtures/llcresi/DpgClient.java:134]: Why is this here?

Please be careful on reviewing the output of the swaggers, they serve this exact purpose.

**Review by weidongxu-microsoft — COMMENTED** [review]
Overall LGTM.

Still some problems commented.

File: protocol-tests/swagger/required-optional-body.json (hunk 3 above)
Comment (haolingdong-msft) [reply protocol-tests/swagger/required-optional-body.json:1]: It just contains different cases in one swagger json so that it's easier to validate the result. But autorest.testserver covers those cases as well in different swagger json. We can remove this json.

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java [hunk 5]
```diff
@@ -43,8 +43,9 @@ protected static void generateProtocolMethodJavadoc(ClientMethod clientMethod, J
                     // ignore if synthesized by modelerfour, i.e. api-version
                     .filter(p -> p.getOrigin() == ParameterSynthesizedOrigin.NONE)
                     .collect(Collectors.toList());
-            if (!queryParameters.isEmpty()) {
+            if (!queryParameters.isEmpty() && hasParametersToPrintInJavadoc(queryParameters)) {
                 optionalParametersJavadoc("Query Parameters", queryParameters, commentBlock);
+                commentBlock.line("You can add these to a request with {@code new RequestOptions.addQueryParam()}");
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java:48]: Recommend to use `{@link RequestOptions#addQueryParam}`, as not every parameter require a `new RequestOptions`

**Comment by haolingdong-msft** [comment]
@srnagar Could you please take a look at this pr?

**Review by weidongxu-microsoft — APPROVED** [review]
